beam/src/Rhyolite/Task/Beam/Worker.hs

```diff
@@ -0,0 +1,151 @@
+{-# Language FlexibleContexts #-}
+{-# Language RankNTypes #-}
+{-# Language ScopedTypeVariables #-}
+{-# Language TypeApplications #-}
+{-# Language TypeFamilies #-}
+-- | Utility functions for creating worker threads, and Beam specific task workers for running SQL.
+module Rhyolite.Task.Beam.Worker where
+
+import Control.Concurrent
+import Control.Concurrent.Async
+import Control.Concurrent.Thread.Delay
+import Control.Exception.Lifted (bracket)
+import Control.Lens ((^.), Lens')
+import Control.Monad (forM)
+import Control.Monad.Cont
+import Control.Monad.Logger (askLoggerIO, MonadLoggerIO)
+import Control.Monad.Logger.Extras (Logger(..))
+import Control.Monad.Trans.Control
+import Data.Time
+import Data.Text (Text)
+
+import Database.Beam
+import Database.Beam.Backend.SQL
+import Database.Beam.Postgres
+import Database.Beam.Postgres.Full
+import Database.Beam.Postgres.Syntax
+import Database.Beam.Query.Internal (QNested)
+import Database.Beam.Schema.Tables
+
+import Database.PostgreSQL.Serializable
+import Rhyolite.DB.Beam
+import Rhyolite.Task.Beam
+
+-- | This function retrieves the first unclaimed task from the table (passed as an argument).
+-- It then assigns this task to itself, and runs the action passed to it.
+-- Once it has the output of that action, it sets the output as the result of the task,
+-- thereby marking it as completed.
+taskWorker ::
+  forall m be db f table a b.
+  ( MonadIO m, MonadLoggerIO m, Database be db
+  , Beamable table, Table table
+  , FromBackendRow be (PrimaryKey table Identity)
+  , FieldsFulfillConstraint (HasSqlEqualityCheck be) (PrimaryKey table)
+  , FieldsFulfillConstraint (HasSqlValueSyntax PgValueSyntax) (PrimaryKey table)
+  , FromBackendRow be a, HasSqlValueSyntax PgValueSyntax b
+  , be ~ Postgres, f ~ QExpr be (QNested (QNested QBaseScope)))
+  => Connection -- ^ Connection to the database
+  -> DatabaseEntity be db (TableEntity table) -- ^ A table containing embedded tasks.
+  -> f Bool -- ^ A filter for selecting tasks from the table.
+  -> (forall x. Lens' (table x) (Task a b x))
+  -> (a -> Serializable (m (Serializable b))) -- ^ The action, whose output is set as the result of the task.
+  -> Text -- ^ Name of the worker
+  -> m Bool
+taskWorker dbConn table ready field go workerName = do
+  logger <- askLoggerIO
+  checkedOutValue <-
+    -- BEGIN Transaction
+
+    -- Do the following inside a transaction:
+    -- 1. Get the first task that is not currently checked out by any worker, lock this task
+    -- 2. Update this task to reflect that it has been checked out by current worker
+    -- 3. Perform the serializable task
+    withTransactionSerializableRunBeamPostgres dbConn $ do
+      primaryKeyAndInput <- runSelectReturningOne $ select $
```

Comment:
Is there a reason not to use `UPDATE...RETURNING`?

Reply:
@ali-abrar Yes, we cannot use [update..returning](https://hackage.haskell.org/package/beam-core-0.9.1.0/docs/Database-Beam-Backend-SQL-BeamExtensions.html) since it returns all the rows that were updated by the update statement.

The reason is that we want to update just one row, and we have no way of identifying that row, we retrieve the set of rows that represent unclaimed tasks, and then limit the results to one.

We could use an update-select query, but I couldn't find a way to do that in beam. I did find support for [CTEs](https://haskell-beam.github.io/beam/user-guide/queries/common-table-expressions/), but I think this is limited to select statements.

There is a variant of insert called [insertFrom](https://hackage.haskell.org/package/beam-core-0.9.1.0/docs/Database-Beam-Query.html#g:35), but I could not find a similar variant for update.

Let me know if I missed something.

Reply:
I thought the second function parameter of `update` can do a subselect? Jank, but I guess one could do a self join of sorts with that with the limit 1? Quite rusty on this stuff, however, so take that with a big pinch of salt.

https://dba.stackexchange.com/questions/69471/postgres-update-limit-1 had a very nicely written answer. I had no idea avoiding the CTE could help with currency issues, but hey that's a nice silver lining.

Reply:
@Ericson2314 I looked at the types, and `update` function cannot accept the `Q` monad, which is used by `select` statements.

I too had encountered the stackexchange link, and I think in our case concurrency is not going to an issue, since we are already inside a serializable transaction.

I'm marking this conversation as resolved for the time being, feel free to mark it as resolved in case there is a better solution.